Sources/Basics/AuthorizationProvider.swift

```diff
@@ -1,19 +1,37 @@
 /*
  This source file is part of the Swift.org open source project
+
  Copyright (c) 2021 Apple Inc. and the Swift project authors
  Licensed under Apache License v2.0 with Runtime Library Exception
+
  See http://swift.org/LICENSE.txt for license information
  See http://swift.org/CONTRIBUTORS.txt for Swift project authors
  */
 
-import Foundation
+import struct Foundation.Data
+import struct Foundation.URL
+#if os(macOS)
+import Security
+#endif
+
+import TSCBasic
+import TSCUtility
 
 public protocol AuthorizationProvider {
-    func authentication(for url: URL) -> (user: String, password: String)?
+    mutating func addOrUpdate(for url: Foundation.URL, user: String, password: String, callback: @escaping (Result<Void, Error>) -> Void)
+    
+    func authentication(for url: Foundation.URL) -> (user: String, password: String)?
+}
+
+public enum AuthorizationProviderError: Error {
+    case noURLHost
+    case notFound
+    case unexpectedPasswordData
```

Comment:
I wasn't sure what this meant until seeing this used in the context of `data(encoding: .utf8)`. What do you think about `invalidPassword` or `cannotEncodePassword`?

Reply:
Renaming to `cannotEncodePassword`